packages/local_auth/local_auth_android/android/src/main/java/io/flutter/plugins/localauth/LocalAuthPlugin.java

```diff
@@ -147,79 +146,78 @@ private void authenticate(MethodCall call, final Result result) {
     }
 
     authInProgress.set(true);
-    AuthCompletionHandler completionHandler =
-        new AuthCompletionHandler() {
-          @Override
-          public void onSuccess() {
-            authenticateSuccess(result);
-          }
-
-          @Override
-          public void onFailure() {
-            authenticateFail(result);
-          }
-
-          @Override
-          public void onError(String code, String error) {
-            if (authInProgress.compareAndSet(true, false)) {
-              result.error(code, error, null);
-            }
-          }
-        };
+    AuthCompletionHandler completionHandler = createAuthCompletionHandler(result);
 
-    // if is biometricOnly try biometric prompt - might not work
     boolean isBiometricOnly = call.argument("biometricOnly");
-    if (isBiometricOnly) {
-      if (!canAuthenticateWithBiometrics()) {
-        if (!hasBiometricHardware()) {
-          completionHandler.onError("NoHardware", "No biometric hardware found");
-        }
-        completionHandler.onError("NotEnrolled", "No biometrics enrolled on this device.");
-        return;
-      }
-      authHelper =
-          new AuthenticationHelper(
-              lifecycle, (FragmentActivity) activity, call, completionHandler, false);
-      authHelper.authenticate();
+    boolean allowCredentials = !isBiometricOnly && canAuthenticateWithDeviceCredential();
+
+    if (canAuthenticateWithBiometrics() || allowCredentials) {
```

Comment:
I haven't tried this out, so maybe I'm just misunderstanding the docs, but do we even need to do this check and the `errorCode` computation below? If we just call `authenticate`, wouldn't we get an immediate error callback for these cases (e.g., we tried to auth with biometrics only, but there were no biometrics) with the right code, which we could then translate into our own error result? It seems like we still may be duplicating some `BiometricPrompt` logic here.

Reply:
Ah, I see. I guess the idea before was that we could catch any issues before trying to authenticate (I assume the old APIs required this), but you're right. Now, we can just send the request and the `AuthenticationHelper` should catch it since it provides the callbacks for authentication requests.

On this note, in a follow-up PR I will add tests for `AuthenticationHelper` since that will do all of the error handling and is currently not tested.